thorlcr/graph/thgraphmaster.cpp

```diff
@@ -1609,44 +1612,43 @@ void CJobMaster::broadcast(ICommunicator &comm, CMessageBuffer &msg, mptag_t mpt
         replyTag = queryJobChannel(0).queryMPServer().createReplyTag();
         msg.setReplyTag(replyTag);
     }
-    if (globals->getPropBool("@broadcastSendAsync", true)) // only here in case of problems/debugging.
+    auto f = [&](unsigned r) -> bool
     {
-        class CSendAsyncfor : public CAsyncFor
+        if (!aborting || comm.verifyConnection(r, (unsigned)-1, false))
         {
-            CMessageBuffer &msg;
-            mptag_t mptag;
-            unsigned timeout;
-            StringAttr errorMsg;
-            ICommunicator &comm;
-        public:
-            CSendAsyncfor(ICommunicator &_comm, CMessageBuffer &_msg, mptag_t _mptag, unsigned _timeout, const char *_errorMsg)
-                : comm(_comm), msg(_msg), mptag(_mptag), timeout(_timeout), errorMsg(_errorMsg)
-            {
-            }
-            void Do(unsigned i)
-            {
-                if (!comm.send(msg, i+1, mptag, timeout))
-                    throw createBCastException(i+1, errorMsg);
-            }
-        } afor(comm, msg, mptag, timeout, errorMsg);
+            if (!comm.send(msg, r, mptag, timeout))
+                throw createBCastException(r, errorMsg);
+            return true;
+        }
+        return false;
+    };
+    std::vector<std::future<bool>> sendResultsFuture;
+    Owned<IException> sendExcept;
+    for (unsigned dstrank=1; dstrank<=groupSizeExcludingMaster; ++dstrank)
+        sendResultsFuture.push_back(std::async(f, dstrank));
+    unsigned numberMsgSent = 0;
+    for (auto & sendResultFuture: sendResultsFuture)
+    {
         try
         {
-            afor.For(groupSizeExcludingMaster, groupSizeExcludingMaster);
+            if (sendResultFuture.get())
+                ++numberMsgSent;
         }
         catch (IException *e)
         {
-            EXCLOG(e, "broadcastSendAsync");
-            abort(e);
-            throw;
+            if (!sendExcept)
```

Comment:
information:  there is a helper function you can use for this (although you may not prefer it):

sendExcept.setownIfNull(e);